rs/nervous_system/agent/src/agent_impl.rs

```diff
@@ -34,23 +34,29 @@ impl CallCanisters for Agent {
         request: R,
     ) -> Result<R::Response, Self::Error> {
         let canister_id = canister_id.into();
-        let request_bytes = request.payload().map_err(AgentCallError::CandidEncode)?;
+        let method = request.method();
+        let payload = request.payload().map_err(AgentCallError::CandidEncode)?;
+
         let response = if request.update() {
-            let request = self
-                .update(&canister_id, request.method())
-                .with_arg(request_bytes)
-                .call()
-                .await?;
-            let (response, _cert) = match request {
+            let call = if let Some(effective_canister_id) = request.effective_canister_id() {
+                self.update(&canister_id, method)
+                    .with_effective_canister_id(effective_canister_id)
+            } else {
+                self.update(&canister_id, method)
+            };
+
+            let response = call.with_arg(payload).call().await?;
+
+            let (response, _cert) = match response {
                 ic_agent::agent::CallResponse::Response(response) => response,
                 ic_agent::agent::CallResponse::Poll(request_id) => {
                     self.wait(&request_id, canister_id).await?
                 }
             };
             response
         } else {
-            self.query(&canister_id, request.method())
-                .with_arg(request_bytes)
+            self.query(&canister_id, method)
```

Comment:
There's also a query call method on the management canister (`fetch_canister_logs`) so maybe we should also include the effective canister ID handling here.

Reply:
Done